### PR TITLE
gather, gather_nd

### PR DESCRIPTION
**Pull request by silee2**
Mostly like tf.gather and tf.gather_nd except for negative index and out of bounds index handling
Added reference implementation. 
Updated Interpreter, CPU (DEX/Codegen) and Generic CPU to use reference implementation.
UT added (including type_prop).
UT is skipped for GPU, iGPU, PlaidML as they don't support the ops yet.

Limitations:
- No autodiff support
- No backdrop (scatter) support
- No documentation 
- No python wrappers
These limitations shout be addressed in future PRs

### CODE REVIEW

File: src/ngraph/op/gather.hpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+#pragma once
+
+#include "ngraph/op/op.hpp"
+
+namespace ngraph
+{
+    namespace op
+    {
+        class Gather : public Op
```
Comment (mbrookhart) [inline src/ngraph/op/gather.hpp]: Docstring?

File: src/ngraph/op/gather_nd.hpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+#pragma once
+
+#include "ngraph/op/op.hpp"
+
+namespace ngraph
+{
+    namespace op
+    {
+        class GatherND : public Op
```
Comment (mbrookhart) [inline src/ngraph/op/gather_nd.hpp]: Docstring

File: src/ngraph/op/gather.hpp (hunk 1 above)
Comment (diyessi) [reply src/ngraph/op/gather.hpp]: I'll be providing a separate .rst PR for doc.

File: src/ngraph/op/gather.cpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+static int PARAMS = 0;
+static int INDICES = 1;
+
+shared_ptr<Node> op::Gather::copy_with_new_args(const NodeVector& new_args) const
+{
+    check_new_args_count(this, new_args);
+    return make_shared<Gather>(new_args.at(PARAMS), new_args.at(INDICES), m_axis);
+}
+
+void op::Gather::validate_and_infer_types()
```
Comment (aprocter) [inline src/ngraph/op/gather.cpp]: I think there will need to be some element type restriction on `INDICES`, right?

File: src/ngraph/op/gather.hpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+namespace ngraph
+{
+    namespace op
+    {
+        /// \brief Gather slices from axis of params according to indices
+        class Gather : public Op
+        {
+        public:
+            // \param params The tensor from which slices are gathered
+            // \param indices Index tensor: Data type be int32_t or int64_t
```
Comment (aprocter) [inline src/ngraph/op/gather.hpp]:
```suggestion
            // \param indices Index tensor: Data type must be `element::i32` or `element::i64`
```

File: test/type_prop.cpp [hunk 5]
```diff
@@ -13084,3 +13084,147 @@ TEST(type_prop, prelu)
     ASSERT_EQ(prelu->get_element_type(), element::f32);
     ASSERT_EQ(prelu->get_shape(), prelu_shape);
 }
+
```
Comment (aprocter) [inline test/type_prop.cpp]: Can we get some tests for failing cases as well?

File: src/ngraph/op/gather_nd.hpp [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+namespace ngraph
+{
+    namespace op
+    {
+        /// \brief Gather slices from params with shapes given by indices
+        class GatherND : public Op
+        {
+        public:
+            // \param params The tensor from which slices are gathered
+            // \param indices Index tensor: Data type be int32_t or int64_t
```
Comment (aprocter) [inline src/ngraph/op/gather_nd.hpp]:
```suggestion
            // \param indices Index tensor: Data type must be `element::i32` or `element::i64`
```

File: src/ngraph/op/gather.hpp [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            // \param axis Axis in params to gather
+            Gather(const std::shared_ptr<Node>& params,
+                   const std::shared_ptr<Node>& indices,
+                   size_t axis = 0)
```
Comment (aprocter) [inline src/ngraph/op/gather.hpp]: I think we should drop the default value here.

File: src/ngraph/op/gather_nd.cpp [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+static int PARAMS = 0;
+static int INDICES = 1;
+
+shared_ptr<Node> op::GatherND::copy_with_new_args(const NodeVector& new_args) const
+{
+    check_new_args_count(this, new_args);
+    return make_shared<GatherND>(new_args.at(PARAMS), new_args.at(INDICES));
+}
+
+void op::GatherND::validate_and_infer_types()
```
Comment (aprocter) [inline src/ngraph/op/gather_nd.cpp]: Ditto the element type restriction on `INDICES`

File: src/ngraph/op/gather.hpp (hunk 7 above)
Comment (aprocter) [reply src/ngraph/op/gather.hpp]: (Up to you, though. No strong feelings.)

File: src/ngraph/op/gather_nd.cpp [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    element::Type result_et = get_input_element_type(PARAMS);
+
+    const PartialShape& params_shape = get_input_partial_shape(PARAMS);
+    const PartialShape& indices_shape = get_input_partial_shape(INDICES);
+
+    NODE_VALIDATION_CHECK(this,
+                          indices_shape.rank().is_dynamic() ||
+                              static_cast<size_t>(indices_shape.rank()) >= 2,
+                          "indices rank is expected to be at least 2");
+
```
Comment (jiminha) [inline src/ngraph/op/gather_nd.cpp]: why the rank is limited to at least 2? tf.gather_nd supports indices of rank size of 1. Is the plan to use gather op instead if the rank size is 1?

File: test/backend_gather.in.cpp [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+#include "util/ndarray.hpp"
+#include "util/random.hpp"
+#include "util/test_control.hpp"
+#include "util/test_tools.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+static string s_manifest = "${MANIFEST}";
+
```
Comment (jiminha) [inline test/backend_gather.in.cpp]: do you have test case for gather with scalar index?

**Pull request merged**